File: haskell/glue/src/Glue/Lib/Math/Logarithmic/Lg.hs
module Glue.Lib.Math.Logarithmic.Lg where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Common logarithm function (lg, log base 10)
-- Mirrors Haskell Glue.Lib.Math.Logarithmic.Lg.lg exactly
lg :: IR Eval
lg = NativeFunc lgImpl

-- Common logarithm function implementation
-- Mirrors Haskell Glue.Lib.Math.Logarithmic.Lg.lgImpl exactly
lgImpl :: IR Eval -> Eval (IR Eval)
lgImpl arg = case arg of
    Integer n -> pure $ Float (logBase 10 (fromIntegral n))
    Float n -> pure $ Float (logBase 10 n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/logarithmic/lg.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Logarithm base 10 function
/// Mirrors Haskell Glue.Lib.Math.Logarithmic.Lg.lg exactly
final Ir lg = IrNativeFunc(lgImpl);

/// Logarithm base 10 function implementation
/// Mirrors Haskell Glue.Lib.Math.Logarithmic.Lg.lgImpl exactly
Eval<Ir> lgImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(
      IrFloat(math.log(n.toDouble()) / math.ln10),
    ),
    IrFloat(value: final n) => Eval.pure(IrFloat(math.log(n) / math.ln10)),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
